polynote-spark/src/main/scala/polynote/kernel/lang/sql/SparkSqlInterpreter.scala

```diff
@@ -131,13 +136,16 @@ class SparkSqlInterpreter(val symbolTable: RuntimeSymbolTable) extends LanguageI
 
   override def shutdown(): IO[Unit] = IO.unit
 
-  private lazy val databaseNames = spark.catalog.listDatabases().map(_.name).collect()
+  private val sessionCatalog = SessionStateThief(spark).catalog
+
+  private var databaseNames = sessionCatalog.listDatabases().toArray
```

Comment:
Even after solving the `ClassCastException`, it turned out that using the `spark.catalog` APIs is just totally broken. It internally calls this same API, which is super fast and great. But then it wraps the result into a `Dataset`, basically requiring executors to come up just so it can send them data it already has just so I can collect it back to the driver. It's nuts.

So `F[_]` that noise... I'm just going right to the source instead, which is fast and good. Uses an unstable package-private API, but that's better than waiting for several minutes for the SQL interpreter to boot because for no apparent reason it needs to allocate executors. And, it tries to allocate a TON of executors, because this list can be pretty large.

Reply:
ouch! 

Not sure I understand what you mean by `IO` that noise though 😉